README.rst
.. _libcamiface:

************************************
libcamiface - camera interface C API
************************************

Overview
========

libcamiface ("camera interface") is a C API that provides a camera and OS
independent image acquisition framework.

There is also a Python wrapper (:mod:`cam_iface`) of the libcamiface
libraries.

.. _libcamiface-supported-cameras:

What cameras are supported?
===========================

Camera support is determined by a libcamiface "backend". Each backend
is a piece of code that creates the libcamiface interface for each
particular camera driver.

As seen below, the supported camera drivers are currently libdc1394,
the Prosilia Gigabit Ethernet SDK, Point Grey Research FlyCapture2,
and the QuickTime SequenceGrabber. For a list of cameras supporting
the libdc1394 software, see
http://damien.douxchamps.net/ieee1394/cameras/. For the Prosilica
Gigabit cameras, see http://www.prosilica.com/. For the Point Grey
cameras, see http://www.ptgrey.com. The QuickTime SequenceGrabber
supports any camera available through QuickTime. This includes the
built-in cameras on Mac laptops and desktops.

.. _libcamiface-supported-rates:

What frame rates, image sizes, bit depths are possible?
=======================================================

This depends primarily on the camera and camera interface. A primary
goal of the Motmot suite is not to limit the capabilities
possible. The fastest frame rates that we routinely use are 500 fps
with a Basler A602f camera using a small region of interest and 200
fps on a Prosilica GE 680 at full frame resolution. The largest image
size we routinely use is 1600x1200 on a Pt. Grey Scorpion camera. The
largest bit depths we routinely use are 12 bits/pixel on a Basler
A622f camera.

Backend status
==============

A number of backends are supported.

.. list-table::
  :header-rows: 1

  * - Backend
    - GNU/Linux i386
    - GNU/Linux x86_64 (amd64)
    - win32 (XP)
    - Mac OS X
  * - libdc1394_
    - |works|
    - |works|
    - |orange| newest library version supports Windows, but untested with libcamiface
    - |mostly works| triggering options disabled
  * - `Prosilica GigE Vision`_
    - |works|
    - |works|
    - |works|
    - |works|
  * - `Point Grey Research FlyCapture2 <http://www.ptgrey.com/products/pgrflycapture/index.asp>`_
    - |orange| untested
    - |orange| untested
    - |works|
    - |NA|
  * - `ImperX`_
    - |NA|
    - |NA|
    - |orange| rudiments present in git 'cruft' branch
    - |NA|
  * - `Basler BCAM 1.8`_
    - |NA|
    - |NA|
    - |orange| rudiments present in git 'cruft' branch, frequent BSOD
    - |NA|
  * - `QuickTime SequenceGrabber`_
    - |NA|
    - |NA|
    - |NA| (supported by driver?)
    - |mostly works| basic functionality works

Key to the above symbols:

* |works| Works well, no known bugs or missing features
* |mostly works| Mostly works, some missing features
* |orange| Not working, although with some effort, could presumably be
   made to work
* |NA| The upstream driver does not support this configuration

.. _libdc1394: http://damien.douxchamps.net/ieee1394/libdc1394/
.. _Prosilica GigE Vision: http://www.prosilica.com
.. _ImperX: http://www.imperx.com/
.. _Basler BCAM 1.8: http://www.baslerweb.com/indizes/beitrag_index_en_21486.html
.. _QuickTime SequenceGrabber: http://developer.apple.com/quicktime/

.. |works| image:: _static/greenlight.png
  :alt: works
  :width: 22
  :height: 22
.. |mostly works| image:: _static/yellowgreenlight.png
  :alt: mostly works
  :width: 22
  :height: 22
.. |orange| image:: _static/redlight.png
  :alt: caution
  :width: 22
  :height: 22
.. |NA| replace:: NA


Download
========

.. Also keep motmot/doc/source/download.rst in sync with download page.

Download official releases from `the download page`__.

__ http://code.astraw.com/libcamiface

Install from binary releases
============================

Mac OS X
--------

The Mac installer is called ``libcamiface-x.y.z-Darwin.dmg``.

This will install the files::

  /usr/include/cam_iface.h
  /usr/bin/liveview-glut-*
  /usr/bin/ ( other demos )
  /usr/lib/libcam_iface_*

To run a demo program, open ``/usr/bin/liveview-glut-mega``.

Windows
-------

The Windows installer is called ``libcamiface-x.y.z-win32.exe``.

This will install the files::

  C:\Program Files\libcamiface x.y.z\bin\simple-prosilica_gige.exe
  C:\Program Files\libcamiface x.y.z\bin\liveview-glut-prosilica_gige.exe
  C:\Program Files\libcamiface x.y.z\include\cam_iface.h
  C:\Program Files\libcamiface x.y.z\lib\cam_iface_prosilica_gige.lib
  C:\Program Files\libcamiface x.y.z\bin\cam_iface_prosilica_gige.dll

To run a demo program, open ``C:\Program Files\libcamiface x.y.z\bin\liveview-glut-prosilica_gige.exe``.

Compile from source
===================

On all platforms, you need to install cmake. cmake is available from
http://www.cmake.org/

You will also need the libraries for any camera software. Cmake should
automatically find these if they are installed in the default
locations.

linux
-----

::

  mkdir build
  cd build
  cmake ..
  make
  make install

To build with debug symbols, include the argument
``-DCMAKE_BUILD_TYPE=Debug`` in your call to cmake. To install in
/usr, include ``-DCMAKE_INSTALL_PREFIX=/usr``. To make verbose
makefiles, include ``-DCMAKE_VERBOSE_MAKEFILE=1``.

To cut a source release::

  VERSION="0.8.0"
  git archive --prefix=libcamiface-$VERSION/ release/$VERSION | gzip -9 > ../libcamiface-$VERSION.tar.gz
  git archive --prefix=libcamiface-$VERSION/ --format=zip release/$VERSION > ../libcamiface-$VERSION.zip

To make a Debian source package::

  VERSION="0.8.0"
  ln ../libcamiface-$VERSION.tar.gz ../libcamiface_$VERSION.orig.tar.gz
  rm -rf ../libcamiface_*.orig.tar.gz.tmp-nest
  git-buildpackage --git-debian-branch=debian --git-upstream-branch=master --git-no-create-orig --git-tarball-dir=.. --git-ignore-new --git-verbose -rfakeroot -S

Mac OS X
--------

Download and install Apple's XCode. This requires signing up (free) as
an Apple ADC member.

::

  mkdir build
  cd build
  cmake ..
  make
  cpack

To build with debug symbols, include the argument
``-DCMAKE_BUILD_TYPE=Debug`` in your call to cmake.

In fact, I use the following commands to set various environment
variables prior to my call to cmake.::

  # You will doubtless need to change these to match your system
  export PROSILICA_CMAKE_DEBUG=1
  export PROSILICA_TEST_LIB_PATHS=/Prosilica\ GigE\ SDK/lib-pc/x86/4.0
  export GLEW_ROOT="/Users/astraw/other-peoples-src/glew/glew-1.5.1"

This will build a Mac installer, called ``libcamiface-x.y.z-Darwin.dmg``.

To build an Xcode project, run cmake with the argument
``-DCMAKE_GENERATOR=Xcode``.

Windows
-------

Windows XP 32bit with CMake 2.6
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Install Microsoft's Visual Studio 2008. (Tested with Express Edition.)
Install CMake.

Open a Visual Studio Command Prompt from Start Menu->All
Programs->Microsoft Visual C++ 2008 Express Edition->Visual Studio
Tools->Visual Studio 2008 Command Prompt. Change directories into the
libcamiface source directory.

Compile in Release mode. (On linux, this would be done with adding
``-DCMAKE_BUILD_TYPE=Release`` in your call to cmake. As I write this,
I'm not sure of the best way to do it in Windows.)

::

  cmakesetup
  rem  In the cmakesetup GUI, set your source and build directories.
  rem  Click "configure".
  rem  In the "Select Generator" menu that pops up, press "NMake Makefiles".
  rem  After it's done configuring, click "configure" again.
  rem  Finally, click "OK".

  rem Now change into your build directory.
  cd build
  nmake

  rem Now, to build an NSIS .exe Windows installer.
  cpack

This will build a Windows installer, called
``libcamiface-x.y.z-win32.exe``.

Windows 7 64bit with CMake 2.8 to make 32 bit libcamiface
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Install Microsoft's Visual Studio 2008. Install CMake.

Open a Visual Studio Command Prompt from Start Menu->All
Programs->Microsoft Visual Studio 2008->Visual Studio
Tools->Visual Studio 2008 Command Prompt. Change directories into the
libcamiface source directory.

::

  cd build
  "C:\Program Files (x86)\CMake 2.8\bin\cmake.exe" .. -G "NMake Makefiles"
  nmake

  rem Now, to build an NSIS .exe Windows installer.
  cpack

This will build a Windows installer, called
``libcamiface-x.y.z-win32.exe``.


Backend notes
=============

prosilica_gige
--------------

Here is an example of setting attributes on the camera using
Prosilica's command line tools::

  export CAM_IP=192.168.1.63
  CamAttr -i $CAM_IP -s StreamBytesPerSecond 123963084
  CamAttr -i $CAM_IP -s PacketSize 1500

Environment variables:

  * *PROSILICA_BACKEND_DEBUG* print various debuging information.

libdc1394
---------

Environment variables:

 * *DC1394_BACKEND_DEBUG* print libdc1394 error messages. (You may
   also be interested in libdc1394's own *DC1394_DEBUG* environment
   variable, which prints debug messages.)

 * *DC1394_BACKEND_1394B* attempt to force use of firewire
    800. (Otherwise defaults to 400.)

 * *DC1394_BACKEND_AUTO_DEBAYER* use dc1394 to de-Bayer the images,
    resulting in RGB8 images (rather than MONO8 Bayer images).

Basler Pylon
------------

Environment variables for CMake:

 * *BASLER_PYLON_LIB_PATH* location of Basler libraries
    (e.g. ``export BASLER_PYLON_LIB_PATH="/opt/pylon/lib64/"``)

 * *BASLER_PYLON_INCLUDE_PATH* location of Basler include files
    (e.g. ``export BASLER_PYLON_INCLUDE_PATH="/opt/pylon/include/"``)

Environment variables for Basler Pylon libraries:

 * *PYLON_ROOT* location of Pylon SDK root (e.g. 
   ``export PYLON_ROOT=/opt/pylon``)


debian/changelog
libcamiface (0.8.0-0ads1) precise; urgency=low

  * New upstream release

 -- Andrew Straw <strawman@astraw.com>  Fri, 08 Jun 2012 17:03:59 +0200

libcamiface (0.7.3-0ads1) precise; urgency=low

  * New upstream release

 -- Andrew Straw <strawman@astraw.com>  Wed, 30 May 2012 15:45:54 +0200

libcamiface (0.7.0-0ads1) precise; urgency=low

  * New upstream release

 -- Andrew Straw <strawman@astraw.com>  Thu, 18 Aug 2011 10:13:49 +0200

libcamiface (0.6.3-0ads1) lucid; urgency=low

  * New upstream release

 -- Andrew Straw <strawman@astraw.com>  Thu, 18 Aug 2011 10:13:49 +0200

libcamiface (0.6.2-0ads2) lucid; urgency=low

  * Actually package new upstream release

 -- Andrew Straw <strawman@astraw.com>  Thu, 24 Jun 2010 16:50:32 -0700

libcamiface (0.6.2-0ads1) lucid; urgency=low

  * New upstream release

 -- Andrew Straw <strawman@astraw.com>  Thu, 24 Jun 2010 16:35:05 -0700

libcamiface (0.6.1-0ads1) lucid; urgency=low

  * New upstream release

 -- Andrew Straw <strawman@astraw.com>  Tue, 22 Jun 2010 10:39:18 -0700

libcamiface (0.5.9-0ads1) hardy; urgency=low

  * New upstream release

 -- Andrew Straw <strawman@astraw.com>  Sun, 24 Jan 2010 23:02:31 -0800

libcamiface (0.5.8-0ads2) hardy; urgency=low

  * Move Prosilica drivers to their own package

 -- Andrew Straw <strawman@astraw.com>  Sat, 07 Nov 2009 16:29:19 -0800

libcamiface (0.5.8-0ads1) hardy; urgency=low

  * New upstream release

 -- Andrew Straw <strawman@astraw.com>  Thu, 05 Nov 2009 09:29:31 -0800

libcamiface (0.5.7-0ads1) hardy; urgency=low

  * New upstream release

 -- Andrew Straw <strawman@astraw.com>  Mon, 02 Nov 2009 10:43:21 -0800

libcamiface (0.5.6-0ads1) hardy; urgency=low

  * New upstream release

 -- Andrew Straw <strawman@astraw.com>  Sun, 01 Nov 2009 11:16:57 -0800

libcamiface (0.5.5-0ads3) hardy; urgency=low

  * Install more Pt. Grey camera udev rules, change mode to 0664

 -- Andrew Straw <strawman@astraw.com>  Fri, 19 Jun 2009 17:49:11 -0700

libcamiface (0.5.5-0ads2) hardy; urgency=low

  * Put Firefly MV USB camera in plugdev group

 -- Andrew Straw <strawman@astraw.com>  Tue, 16 Jun 2009 11:25:26 -0700

libcamiface (0.5.5-0ads1) hardy; urgency=low

  * New upstream release

 -- Andrew Straw <strawman@astraw.com>  Sat, 13 Jun 2009 18:06:06 -0700

libcamiface (0.5.4-0ads2) hardy; urgency=low

  * Install missing GLSL shader source code
  * Install udev rules with debhelper

 -- Andrew Straw <strawman@astraw.com>  Sat, 13 Jun 2009 15:11:59 -0700

libcamiface (0.5.4-0ads1) hardy; urgency=low

  * New upstream release

 -- Andrew Straw <strawman@astraw.com>  Sat, 13 Jun 2009 12:47:50 -0700

libcamiface (0.5.3-0ads1) hardy; urgency=low

  * New upstream release

 -- Andrew Straw <strawman@astraw.com>  Sat, 02 May 2009 18:45:30 -0700

libcamiface (0.5.2-0ads1) jaunty; urgency=low

  * New upstream release

 -- Andrew Straw <strawman@astraw.com>  Thu, 30 Apr 2009 20:34:05 -0700

libcamiface (0.5.1-0ads2) hardy; urgency=low

  * New upstream version using cmake
  * Switch to debhelper 7

 -- Andrew Straw <strawman@astraw.com>  Wed, 22 Apr 2009 02:08:24 -0700

libcamiface (0.5.1-0ads1) hardy; urgency=low

  * New Upstream Version

 -- Andrew Straw <strawman@astraw.com>  Thu, 09 Apr 2009 18:52:23 -0700

libcamiface (0.5-0ads2) hardy; urgency=low

  * Move prosilica stuff to own package

 -- Andrew Straw <strawman@astraw.com>  Fri, 03 Apr 2009 12:33:54 -0700

libcamiface (0.5-0ads1) hardy; urgency=low

  * New release

 -- Andrew Straw <strawman@astraw.com>  Tue, 31 Mar 2009 11:20:05 -0700

camiface (0.4.dev745-0ads1) jaunty; urgency=low

  * latest updates from svn

 -- Andrew Straw <strawman@astraw.com>  Thu, 19 Mar 2009 16:30:14 -0700

camiface (0.4.dev745-0ads1) jaunty; urgency=low

  * fix FTBFS on jaunty
  * latest updates from svn

 -- Andrew Straw <strawman@astraw.com>  Sat, 14 Mar 2009 21:50:40 -0700

camiface (0.4.dev723-0ads1) hardy; urgency=low

  * latest updates from svn

 -- Andrew Straw <strawman@astraw.com>  Thu, 19 Feb 2009 22:48:38 -0800

camiface (0.4.dev716-0ads1) hardy; urgency=low

  * latest updates from svn

 -- Andrew Straw <strawman@astraw.com>  Fri, 06 Feb 2009 10:58:58 -0800

camiface (0.4.dev681-0ads1) hardy; urgency=low

  * Get trigger mode

 -- Andrew Straw <strawman@astraw.com>  Tue, 09 Dec 2008 10:03:49 -0800

camiface (0.4.dev680-0ads1) hardy; urgency=low

  * Fix segfault when image size changes
  * Experimental support for 1394B with libdc1394

 -- Andrew Straw <strawman@astraw.com>  Tue, 09 Dec 2008 10:03:49 -0800

camiface (0.4.dev650-0ads1) gutsy; urgency=low

  * Fix bug in dc1394 set_frame_offset() call.

 -- Andrew Straw <strawman@astraw.com>  Fri, 17 Oct 2008 12:47:42 -0700

camiface (0.4.dev640-0ads1) gutsy; urgency=low

  * New release


demo/cam2theora.c
/*

Copyright (c) 2004-2009, California Institute of Technology. All
rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are
met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.

    * Redistributions in binary form must reproduce the above
      copyright notice, this list of conditions and the following
      disclaimer in the documentation and/or other materials provided
      with the distribution.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

 */


#include <stdio.h>
#ifdef _WIN32
#include <Windows.h>
#include <sys/timeb.h>
#else
#include <sys/time.h>
#endif
#include <stdlib.h>
#include <time.h>
#include <string.h>
#include "cam_iface.h"

#include <ogg/ogg.h>
#include <theora/theora.h>

/* globals */
static FILE *ogg_fp = NULL;
static int theora_initialized = 0;
static ogg_stream_state ogg_os;
static theora_state theora_td;
static theora_info theora_ti;

double my_floattime() {
#ifdef _WIN32
#if _MSC_VER == 1310
  struct _timeb t;
  _ftime(&t);
  return (double)t.time + (double)t.millitm * (double)0.001;
#else
  struct _timeb t;
  if (_ftime_s(&t)==0) {
    return (double)t.time + (double)t.millitm * (double)0.001;
  }
  else {
    return 0.0;
  }
#endif
#else
  struct timeval t;
  if (gettimeofday(&t, (struct timezone *)NULL) == 0)
    return (double)t.tv_sec + t.tv_usec*0.000001;
  else
    return 0.0;
#endif
}

#define _check_error() {                                                \
    int _check_error_err;                                               \
    _check_error_err = cam_iface_have_error();                          \
    if (_check_error_err != 0) {                                        \
                                                                        \
      fprintf(stderr,"%s:%d %s\n", __FILE__,__LINE__,cam_iface_get_error_string()); \
      exit(1);                                                          \
    }                                                                   \
  }                                                                     \

void show_usage(char * cmd) {
  printf("usage: %s FILENAME [num_frames]\n",cmd);
  printf("  where FILENAME is the name of the .ogv theora file to write\n");
  printf("        num_frames can be a number or 'forever'\n");
  exit(1);
}

static int
theora_open(const char *pathname)
{
  ogg_packet op;
  ogg_page og;
  theora_comment tc;

  ogg_fp = fopen(pathname, "wb");
  if(!ogg_fp) {
    fprintf(stderr, "%s: error: %s\n",
      pathname, "couldn't open output file");
    return 1;
  }

  if(ogg_stream_init(&ogg_os, rand())) {
    fprintf(stderr, "%s: error: %s\n",
      pathname, "couldn't create ogg stream state");
    return 1;
  }

  if(theora_encode_init(&theora_td, &theora_ti)) {
    fprintf(stderr, "%s: error: %s\n",
      pathname, "couldn't initialize theora encoding");
    return 1;
  }

  theora_encode_header(&theora_td, &op);
  ogg_stream_packetin(&ogg_os, &op);
  if(ogg_stream_pageout(&ogg_os, &og)) {
    fwrite(og.header, og.header_len, 1, ogg_fp);
    fwrite(og.body, og.body_len, 1, ogg_fp);
  }

  theora_comment_init(&tc);
  theora_encode_comment(&tc, &op);
  ogg_stream_packetin(&ogg_os, &op);
  if(ogg_stream_pageout(&ogg_os, &og)) {
    fwrite(og.header, og.header_len, 1, ogg_fp);
    fwrite(og.body, og.body_len, 1, ogg_fp);
  }

  theora_encode_tables(&theora_td, &op);
  ogg_stream_packetin(&ogg_os, &op);
  if(ogg_stream_pageout(&ogg_os, &og)) {
    fwrite(og.header, og.header_len, 1, ogg_fp);
    fwrite(og.body, og.body_len, 1, ogg_fp);
  }

  if(ogg_stream_flush(&ogg_os, &og)) {
    fwrite(og.header, og.header_len, 1, ogg_fp);
    fwrite(og.body, og.body_len, 1, ogg_fp);
  }

  return 0;
}

typedef struct
{
  unsigned char R;
  unsigned char G;
  unsigned char B;
}
RGB888_t;

typedef struct
{
  unsigned char Y;
  unsigned char U;
  unsigned char V;
}
YUV444_t;

#define min(x,y)				\
  (x)<(y)?x:y

YUV444_t RGB888toYUV444(unsigned char r, unsigned char g, unsigned char b) {
  // from http://en.wikipedia.org/wiki/YUV
  YUV444_t result;
  result.Y = min(abs(r * 2104 + g * 4130 + b * 802 + 4096 + 131072) >> 13, 235);
  result.U = min(abs(r * -1214 + g * -2384 + b * 3598 + 4096 + 1048576) >> 13, 240);
  result.V = min(abs(r * 3598 + g * -3013 + b * -585 + 4096 + 1048576) >> 13, 240) ;
  return result;
}

char * malloc_and_convert(char *input, CameraPixelCoding coding,
                          size_t input_stride, size_t w, size_t h) {
  char*output;
  size_t output_stride;
  int i,j;

  output_stride = w*3/2;
  output = (char*)malloc(output_stride*h);
  if (output==NULL) {
    return output;
  }

  switch(coding){
  case CAM_IFACE_RGB8:
    fprintf(stderr,"not doing real RGB->YUV422 conversion\n");
    for (i=0; i<h; i++) {
      memcpy(output + (i*output_stride),
             input + (i*input_stride),
             output_stride);
    }
    break;
  default:
    fprintf(stderr,"do not know how to convert image for this format\n");
    free(output);
    return NULL;
    break;
  }

  return output;
}

static int
theora_write_frame(unsigned long w, unsigned long h, unsigned char *yuv)
{
  yuv_buffer yuv_buf;
  ogg_packet op;
  ogg_page og;

  unsigned long yuv_w;
  unsigned long yuv_h;

  unsigned char *yuv_y;
  unsigned char *yuv_u;
  unsigned char *yuv_v;

  unsigned int x;
  unsigned int y;
  
  /* Must hold: yuv_w >= w */
  yuv_w = (w + 15) & ~15;

  /* Must hold: yuv_h >= h */
  yuv_h = (h + 15) & ~15;

  yuv_y = malloc(yuv_w * yuv_h);
  yuv_u = malloc(yuv_w * yuv_h / 4);
  yuv_v = malloc(yuv_w * yuv_h / 4);

  yuv_buf.y_width = yuv_w;
  yuv_buf.y_height = yuv_h;
  yuv_buf.y_stride = yuv_w;
  yuv_buf.uv_width = yuv_w >> 1;
  yuv_buf.uv_height = yuv_h >> 1;
  yuv_buf.uv_stride = yuv_w >> 1;
  yuv_buf.y = yuv_y;
  yuv_buf.u = yuv_u;
  yuv_buf.v = yuv_v;

  for(y = 0; y < yuv_h; y++) {
    for(x = 0; x < yuv_w; x++) {
      yuv_y[x + y * yuv_w] = 0;
    }
  }

  for(y = 0; y < yuv_h; y += 2) {
    for(x = 0; x < yuv_w; x += 2) {
      yuv_u[(x >> 1) + (y >> 1) * (yuv_w >> 1)] = 0;
      yuv_v[(x >> 1) + (y >> 1) * (yuv_w >> 1)] = 0;
    }
  }

  for(y = 0; y < h; y++) {
    for(x = 0; x < w; x++) {
      yuv_y[x + y * yuv_w] = yuv[3 * (x + y * w) + 0];
    }
  }

  for(y = 0; y < h; y += 2) {
    for(x = 0; x < w; x += 2) {
      yuv_u[(x >> 1) + (y >> 1) * (yuv_w >> 1)] =
        yuv[3 * (x + y * w) + 1];
      yuv_v[(x >> 1) + (y >> 1) * (yuv_w >> 1)] =
        yuv[3 * (x + y * w) + 2];
    }
  }

  if(theora_encode_YUVin(&theora_td, &yuv_buf)) {
    fprintf(stderr, "error: could not encode frame\n");
    return 1;
  }

  if(!theora_encode_packetout(&theora_td, 0, &op)) {
    fprintf(stderr, "error: could not read packets\n");
    return 1;
  }

  ogg_stream_packetin(&ogg_os, &op);
  if(ogg_stream_pageout(&ogg_os, &og)) {
    fwrite(og.header, og.header_len, 1, ogg_fp);
    fwrite(og.body, og.body_len, 1, ogg_fp);
  }

  free(yuv_y);
  free(yuv_u);
  free(yuv_v);

  return 0;
}


static void
theora_close(void)
{
  ogg_packet op;
  ogg_page og;

  if (theora_initialized) {
    theora_encode_packetout(&theora_td, 1, &op);
    if(ogg_stream_pageout(&ogg_os, &og)) {
      fwrite(og.header, og.header_len, 1, ogg_fp);
      fwrite(og.body, og.body_len, 1, ogg_fp);
    }
  
    theora_info_clear(&theora_ti);
    theora_clear(&theora_td);
  
    fflush(ogg_fp);
    fclose(ogg_fp);
  }
  
  ogg_stream_clear(&ogg_os);
}

int main(int argc, char** argv) {
  CamContext *cc;
  unsigned char *pixels;

  int device_number,ncams,num_buffers;

  double last_fps_print, now, t_diff;
  double fps;
  int n_frames;
  int buffer_size;
  int num_modes, num_props, num_trigger_modes;
  char mode_string[255];
  int i,mode_number;
  CameraPropertyInfo cam_props;
  long prop_value;
  int prop_auto;
  int errnum;
  int left, top;
  int width, height;
  int do_num_frames;
  cam_iface_constructor_func_t new_CamContext;
  Camwire_id cam_info_struct;
  char *filename;
  unsigned char *yuv;

  cam_iface_startup_with_version_check();
  _check_error();

  if (argc<1) {
    show_usage(argv[0]);
  }

  filename = argv[1];

  if (argc>2) {
    if (strcmp(argv[2],"forever")==0) {
      do_num_frames = -1;
    } else if (sscanf(argv[2],"%d",&do_num_frames)==0) {
      show_usage(argv[0]);
    }
  } else {
    do_num_frames = -1;
  }

  for (i=0;i<argc;i++) {
    printf("%d: %s\n",i,argv[i]);
  }
  printf("using driver %s\n",cam_iface_get_driver_name());

  ncams = cam_iface_get_num_cameras();
  _check_error();

  if (ncams<1) {

    printf("no cameras found, will now exit\n");

    cam_iface_shutdown();
    _check_error();

    exit(1);
  }
  _check_error();

  printf("%d camera(s) found.\n",ncams);
  for (i=0; i<ncams; i++) {
    cam_iface_get_camera_info(i, &cam_info_struct);
    printf("  camera %d:\n",i);
    printf("    vendor: %s\n",cam_info_struct.vendor);
    printf("    model: %s\n",cam_info_struct.model);
    printf("    chip: %s\n",cam_info_struct.chip);
  }

  device_number = ncams-1;

  printf("choosing camera %d\n",device_number);

  cam_iface_get_num_modes(device_number, &num_modes);
  _check_error();

  printf("%d mode(s) available:\n",num_modes);

  mode_number = 0;

  for (i=0; i<num_modes; i++) {
    cam_iface_get_mode_string(device_number,i,mode_string,255);
    if (strstr(mode_string,"FORMAT7_0")!=NULL) {
      if (strstr(mode_string,"MONO8")!=NULL) {
        // pick this mode
        mode_number = i;
      }
    }
    printf("  %d: %s\n",i,mode_string);
  }

  printf("Choosing mode %d\n",mode_number);

  num_buffers = 5;

  new_CamContext = cam_iface_get_constructor_func(device_number);
  cc = new_CamContext(device_number,num_buffers,mode_number);
  _check_error();

  CamContext_get_frame_roi(cc, &left, &top, &width, &height);
  _check_error();

    if(!theora_initialized) {
      theora_info_init(&theora_ti);

      theora_ti.width = ((width + 15) >>4)<<4;
      theora_ti.height = ((height + 15)>>4)<<4;
      theora_ti.frame_width = width;
      theora_ti.frame_height = height;
      theora_ti.offset_x = 0;
      theora_ti.offset_y = 0;
      theora_ti.fps_numerator = 30;
      theora_ti.fps_denominator = 1;
      theora_ti.aspect_numerator = width;
      theora_ti.aspect_denominator = height;
      theora_ti.colorspace = OC_CS_UNSPECIFIED;
      theora_ti.pixelformat = OC_PF_420;
      theora_ti.target_bitrate = 0;
      theora_ti.quality = 40; /* between 0 and 63 */

      theora_ti.dropframes_p = 0;
      theora_ti.quick_p = 1;
      theora_ti.keyframe_auto_p = 1;
      theora_ti.keyframe_frequency = 64;
      theora_ti.keyframe_frequency_force = 64;
      theora_ti.keyframe_data_target_bitrate = 0;
      theora_ti.keyframe_mindistance = 8;
      theora_ti.noise_sensitivity = 1;


      if(theora_open(filename)) {
        cam_iface_shutdown();
        _check_error();

        exit(1);
      }
    }


  CamContext_get_num_framebuffers(cc,&num_buffers);
  printf("allocated %d buffers\n",num_buffers);

  CamContext_get_num_camera_properties(cc,&num_props);
  _check_error();

  for (i=0; i<num_props; i++) {
    CamContext_get_camera_property_info(cc,i,&cam_props);
    _check_error();

    if (strcmp(cam_props.name,"white balance")==0) {
      fprintf(stderr,"WARNING: ignoring white balance property\n");
      continue;
    }

    if (cam_props.is_present) {
      CamContext_get_camera_property(cc,i,&prop_value,&prop_auto);
      _check_error();
      printf("  %s: %ld\n",cam_props.name,prop_value);
    } else {
      printf("  %s: not present\n",cam_props.name);
    }
  }

  CamContext_get_buffer_size(cc,&buffer_size);
  _check_error();

  if (buffer_size == 0) {
    fprintf(stderr,"buffer size was 0 in %s, line %d\n",__FILE__,__LINE__);
    exit(1);
  }

#define USE_COPY
#ifdef USE_COPY
  pixels = (unsigned char *)malloc( buffer_size );
  if (pixels==NULL) {
    fprintf(stderr,"couldn't allocate memory in %s, line %d\n",__FILE__,__LINE__);
    exit(1);
  }
#endif

  CamContext_start_camera(cc);
  _check_error();

  last_fps_print = my_floattime();
  n_frames = 0;

  if (do_num_frames < 0) {
    printf("will now run forever. press Ctrl-C to interrupt\n");
  } else {
    printf("will now grab %d frames.\n",do_num_frames);
  }

  CamContext_get_num_trigger_modes( cc, &num_trigger_modes );
  _check_error();

  printf("trigger modes:\n");
  for (i =0; i<num_trigger_modes; i++) {
    CamContext_get_trigger_mode_string( cc, i, mode_string, 255 );
    printf("  %d: %s\n",i,mode_string);
  }
  printf("\n");

  while (1) {
    if (do_num_frames>=0) {
      do_num_frames--;
      if (do_num_frames<0) break;
    }
#ifdef USE_COPY
    //CamContext_grab_next_frame_blocking(cc,pixels,0.2); // timeout after 200 msec
    CamContext_grab_next_frame_blocking(cc,pixels,-1.0f); // never timeout
    errnum = cam_iface_have_error();
    if (errnum == CAM_IFACE_FRAME_TIMEOUT) {
      cam_iface_clear_error();
      fprintf(stdout,"T");
      fflush(stdout);
      continue; // wait again
    }
    if (errnum == CAM_IFACE_FRAME_DATA_MISSING_ERROR) {
      cam_iface_clear_error();
      fprintf(stdout,"M");
      fflush(stdout);
    } else if (errnum == CAM_IFACE_FRAME_INTERRUPTED_SYSCALL) {
      cam_iface_clear_error();
      fprintf(stdout,"I");
      fflush(stdout);
    } else {
      _check_error();
      fprintf(stdout,".");
      fflush(stdout);
    }
    now = my_floattime();
    n_frames += 1;
#else
    CamContext_point_next_frame_blocking(cc,&pixels,-1.0f);
    now = my_floattime();
    n_frames += 1;
    _check_error();
    fprintf(stdout,".");
    fflush(stdout);
    CamContext_unpoint_frame(cc);
    _check_error();
#endif

    t_diff = now-last_fps_print;
    if (t_diff > 5.0) {
      fps = n_frames/t_diff;
      fprintf(stdout,"%.1f fps\n",fps);
      last_fps_print = now;
      n_frames = 0;
    }

    yuv = malloc_and_convert(pixels,cc->coding,width*3,width,height);
    if (yuv==NULL) {
      theora_close();
      cam_iface_shutdown();
      _check_error();
      exit(1);
    }
    if(theora_write_frame(width, height, yuv)) {
      theora_close();
      cam_iface_shutdown();
      _check_error();
      exit(1);
    }
    free(yuv);

  }
  theora_close();

  printf("\n");
  delete_CamContext(cc);
  _check_error();

  cam_iface_shutdown();
  _check_error();

#ifdef USE_COPY
  free(pixels);
#endif

  return 0;
}
